### PR TITLE
Feature/add yubikey

### PR DESCRIPTION
**Pull request by j0hnsmith**
Adds Yubikey integration, #230.

### CODE REVIEW

**Comment by j0hnsmith** [comment]
The build is failing because `libpcsclite` is now required, I'm not sure if/how I can modify the build env (happy to try if someone can guide me). Pretty sure that's not the only new package needed.

**Comment by ismailyenigul** [comment]
Great job!
I look forward to this feature. I can be first beta tester :)

**Comment by Isaacpm** [comment]
Hi, is there any update on this PR? It would be very useful to have.

**Comment by j0hnsmith** [comment]
If you don't want to wait for this PR to be merged you can build it yourself, see https://github.com/99designs/aws-vault#development (that's what I've done, works well for me, I use it most days). If you do please report here with your experience of the functionality as it will add impetus to this PR.

**Comment by Isaacpm** [comment]
I will give that a try, and report here, thanks!

**Comment by p0bailey** [comment]
Hello there,

Great work! Any updates on when this pr would be merged into master ?

phillip

**Comment by j0hnsmith** [comment]
Something like this should work

```
cd ${GOPATH}/src/github.com
git clone git@github.com:99designs/aws-vault.git
git remote add git@github.com:j0hnsmith/aws-vault.git
git fetch j0hnsmith feature/add_yubikey
git checkout j0hnsmith/feature/add_yubikey

make build
# now you need to sign and install, see https://github.com/99designs/aws-vault#self-signing-your-binary
```

**Comment by StevenACoffman** [comment]
I found the code signing directions lacking.

#### Deal with signing aws-vault

1. Start keychain access
2. Keychain access -> Certificate assistance -> Create a certificate
3. Name it whatever you like (I choose AWSVAULTSIGNER)
4. Check "Let me override defaults"
5. Just click continue until you reach "Extended Key Usage Extension" select "Code Signing"
6. Just click continue until you reach the point to select Keychain, select the *System* keychain
7. type `codesign --force --sign name-of-certificate /Users/${HOME}/go/src/github.com/99designs/aws-vault/aws-vault`  in your console
8. You can then sign your binary like this:

```bash
go get -u -v github.com/99designs/aws-vault
cd $GOPATH/src/github.com/99designs/aws-vault
make build
codesign --force --sign name-of-certificate ${GOPATH}/src/github.com/99designs/aws-vault/aws-vault
export PATH="$(PWD):$PATH"
# verify signature
codesign -dvv $(which aws-vault) 2>&1 | grep Authority
```

### Verify signature
    $ codesign -dvv $(which aws-vault) 2>&1 | grep Authority
Check out Apple's guide on it [here](http://web.archive.org/web/20090119080759/http://developer.apple.com/documentation/Security/Conceptual/CodeSigningGuide/Procedures/chapter_3_section_2.html), or find it in `Keychain Access > Certificate Assistant > Create Certificate > Code Signing Certificate`.

**Comment by viraptor** [comment]
@j0hnsmith I'm getting the same issue. After switching to your branch:
```
➜  aws-vault git:(270eaf0) make build
go build -o aws-vault -ldflags="-X main.Version=v4.4.1-23-g270eaf0 -s -w" .
main.go:6:2: cannot find package "github.com/99designs/aws-vault/cli" in any of:
	/Users/viraptor/go/src/github.com/aws-vault/vendor/github.com/99designs/aws-vault/cli (vendor tree)
	/usr/local/Cellar/go/1.12/libexec/src/github.com/99designs/aws-vault/cli (from $GOROOT)
	/Users/viraptor/go/src/github.com/99designs/aws-vault/cli (from $GOPATH)
make: *** [build] Error 1

➜  aws-vault git:(270eaf0) go get
# github.com/aws-vault
./main.go:27:22: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureGlobals
./main.go:28:25: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureAddCommand
./main.go:29:26: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureListCommand
./main.go:30:28: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureRotateCommand
./main.go:31:26: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureExecCommand
./main.go:32:28: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureRemoveCommand
./main.go:33:27: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureLoginCommand
./main.go:34:28: cannot use app (type *"github.com/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application) as type *"github.com/99designs/aws-vault/vendor/gopkg.in/alecthomas/kingpin.v2".Application in argument to cli.ConfigureServerCommand
./main.go:35:2: undefined: cli.ConfigureAddYubikeyCommand
./main.go:36:2: undefined: cli.ConfigureRemoveYubikeyCommand
./main.go:36:2: too many errors
```

That's using go1.12 on a mac

**Comment by j0hnsmith** [comment]
@viraptor if you follow the instructions in https://github.com/99designs/aws-vault/pull/316#issuecomment-459323832 all should be good.

**Comment by j0hnsmith** [comment]
@mtibben are you happy with the general premise of this PR? If so, what needs to happen to get the build to pass?

**Comment by viraptor** [comment]
@j0hnsmith That's what I did, with some fixes:

```
cd ${GOPATH}/src/github.com
git clone git@github.com:99designs/aws-vault.git
cd aws-vault
git remote add j0hnsmith git@github.com:j0hnsmith/aws-vault.git
git fetch j0hnsmith feature/add_yubikey
git checkout j0hnsmith/feature/add_yubikey
make build
```
But there is some reproducible issue it seems.

**Comment by mtibben** [comment]
@j0hnsmith FYI I've just rebased this

**Comment by j0hnsmith** [comment]
@viraptor I've just built and installed the rebased version using `1.11.5`, worked ok for me, you could try that.

**Comment by viraptor** [comment]
Found the problem - missing directory in the structure - fixed here:
```
mkdir -p ${GOPATH}/src/github.com/99designs
cd ${GOPATH}/src/github.com/99designs
git clone git@github.com:99designs/aws-vault.git
cd aws-vault
git remote add j0hnsmith git@github.com:j0hnsmith/aws-vault.git
git fetch j0hnsmith feature/add_yubikey
git checkout j0hnsmith/feature/add_yubikey
make build
```

**Comment by davxiao** [comment]
Is this already merged to the latest release or it will be included?

**Comment by jottr** [comment]
`libpcsclite` seems to be missing. Could someone please update the dependencies and merge this? 
Lots of people are waiting for this. :)

**Comment by jottr** [comment]
@j0hnsmith I've built and tried using the yubikey with `aws-vault`, but couldn't get it to work: 

I've added the ARN of the yubikey, to `~/.aws/config` with `mfa_serial` as the key. 
I am prompted for a token for the correct mfa ARN when I run `aws-vault exec yubiawsaccount -- aws`, but pressing the button on the key does nothing. 
Not sure what I'm missing

**Comment by jottr** [comment]
@lox @j0hnsmith can this PR please be expedited? 
Having U2F support in `aws-vault` should be considered to be on of the most imporrant features of a tool that facilitates authentication against sensitive server infrastructure.

**Comment by lox** [comment]
What's the status of this @j0hnsmith? I'm reviewing PR's presently, is this ready for review?

**Comment by j0hnsmith** [comment]
@lox I've just rebased against master so yes, absolutely. 

The build is failing because `libpcsclite` needs to be available in the travis environment to build, pretty sure that's not the only missing package. I don't know how to add it, if you can guide me I'm happy to do it, alternatively if you know what needs to be done I'm happy for you to do it.

**Comment by lox** [comment]
Ok cool. I'll see if I can figure out how to get `libpcsclite` installed and I'll leave some notes throughout the PR.

**Comment by asiragusa** [comment]
@j0hnsmith great job! I have tested your PR and I found a flaw in ykoath (https://github.com/yawn/ykoath/pull/8) if you stored tokens that require the --touch and tokens that don't, in some cases this PR won't work.

I did not have a reply for 10 days, but it would be nice if this PR would work properly. Can you update the commit ID in `go.mod`?

Also, I found a small quirk with the MFA device being added even if the tokens used are wrong, in this case you might try to do some cleanup. It's not a major issue, as manually doing `aws-vault remove-yubikey` fixes the problem.

**Comment by asiragusa** [comment]
@j0hnsmith it would be nice also if we could have more feedback once the Yubikey button has been pressed. Today it shows:
```
waiting for yubikey touch...
```

And nothing more once you pressed it. For tasks that don't output something for seconds it's impossible to know if you have correctly pressed the button or not. If you touch the button multiple times, the console is filled with random chars, otherwise it timeouts and asks for a token.

A simple `OK\n` aligned with the `waiting for yubikey touch...` would make it

**Comment by j0hnsmith** [comment]
@asiragusa Thanks for testing and giving feedback. I've added `OK\n` once touch has happened.

>  I found a small quirk with the MFA device being added even if the tokens used are wrong

In what situation would the token be wrong?

> Can you update the commit ID in go.mod?

What do you want it updated to, surely your PR has to be merged first?

**Comment by asiragusa** [comment]
> @asiragusa Thanks for testing and giving feedback. I've added `OK\n` once touch has happened.
Great, thanks!

> 
> > I found a small quirk with the MFA device being added even if the tokens used are wrong
> 
> In what situation would the token be wrong?
It happens if you already have both tokens with and without touch requirement in your Yubikey, so it's mainly an edge case.

> > Can you update the commit ID in go.mod?
> 
> What do you want it updated to, surely your PR has to be merged first?
Given that it's an edge case we can merge this PR and wait for https://github.com/yawn/ykoath/pull/8 to be merged. An issue to remind to update the go.mod will do the trick ;)

**Comment by asiragusa** [comment]
@j0hnsmith I have found a small issue using the `remove-yubikey` command.

In my environment I have forbidden to remove MFA devices without using a MFA token and probably the `remove-yubikey` command doesn't generate one for the operation.

The output I have is the following:

```
$ aws-vault remove-yubikey xxx yyy
Removing yubikey for user xxx using profile yyy)
aws-vault: error: error removing yubikey
%!(EXTRA *errors.withStack=failed to delete virtual AWS MFA device with serial "arn:aws:iam::zzz:mfa/xxx": AccessDenied: User: arn:aws:iam::zzz:user/xxx is not authorized to perform: iam:DeleteVirtualMFADevice on resource: arn:aws:iam::zzz:mfa/xxx
	status code: 403, request id: req-id)
```

**Comment by j0hnsmith** [comment]
@asiragusa I assume your policies require you use multi factor to auth before removing a multi factor device eg `"aws:MultiFactorAuthAge":"28800"`, if you should be able to just `aws-vault exec` then `aws-vault remove-yubikey` from within that session.

If your setup is different please provide an example of the iam user policy in use that's preventing the removal.

**Comment by suominen** [comment]
Adding the libpcsclite-dev package fixes the build: https://github.com/j0hnsmith/aws-vault/pull/1

**Comment by j0hnsmith** [comment]
@lox build passed, how's the review coming along?

File: mfa/iamyubikey.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete virtual AWS MFA device with serial %q", *serial)
+	}
+
+	return nil
+
+}
+
+// CallerIdentityToSerial converts a caller identity ARN to a MFA serial
+func CallerIdentityToSerial(i *string) (string, error) {
```
Comment (lox) [inline mfa/iamyubikey.go]: Does this need to be public?

File: mfa/iamyubikey.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error decoding secret")
+	}
+
+	return res.VirtualMFADevice.SerialNumber, secret, nil
+}
+
+func (m *MFA) enable(username string, serial *string, secret []byte) error {
+
```
Comment (lox) [inline mfa/iamyubikey.go]: Remove space

File: mfa/iamyubikey.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		return err
+	}
+
+	return nil
+
+}
+
+// create creates the virtual MFA device
+func (m *MFA) create(username string) (*string, []byte, error) {
+
```
Comment (lox) [inline mfa/iamyubikey.go]: Remove space

**Comment by lox** [comment]
Approved @j0hnsmith, only made some minor nitpicks on spacing.

**Comment by j0hnsmith** [comment]
@lox changes done

**Comment by asiragusa** [comment]
@j0hnsmith I tried to run an exec, but it didn’t work either:

```bash
$ aws-vault exec <profile> -- aws s3 ls
waiting for yubikey touch...

An error occurred (AccessDenied) when calling the ListBuckets operation: Access Denied # the error doesn't matter here

$ aws-vault remove-yubikey <user> <profile>
Removing yubikey for user <user> using profile <profile>)
aws-vault: error: error removing yubikey
%!(EXTRA *errors.withStack=failed to delete virtual AWS MFA device with serial "arn:aws:iam::<account>:mfa/<user>": AccessDenied: User: arn:aws:iam::<account>:user/<user> is not authorized to perform: iam:DeleteVirtualMFADevice on resource: arn:aws:iam::<account>:mfa/<user>
	status code: 403, request id: xxxx)
```

The user policies are the following:
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": "iam:ChangePassword",
      "Resource": "arn:aws:iam::<account>:user/${aws:username}"
    },
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": "iam:GetAccountPasswordPolicy",
      "Resource": "*"
    },
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": "iam:GetLoginProfile",
      "Resource": "arn:aws:iam::<account>:user/${aws:username}",
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    }
  ]
}
```
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": [
        "iam:UpdateAccessKey",
        "iam:ListAccessKeys",
        "iam:GetUser",
        "iam:GetAccessKeyLastUsed",
        "iam:DeleteAccessKey",
        "iam:CreateAccessKey"
      ],
      "Resource": "arn:aws:iam::<account>:user/${aws:username}"
    }
  ]
}
```
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "AllowUsersToCreateEnableResyncTheirOwnVirtualMFADevice",
      "Effect": "Allow",
      "Action": [
        "iam:ResyncMFADevice",
        "iam:EnableMFADevice",
        "iam:CreateVirtualMFADevice"
      ],
      "Resource": [
        "arn:aws:iam::<account>:user/${aws:username}",
        "arn:aws:iam::<account>:mfa/${aws:username}"
      ]
    },
    {
      "Sid": "AllowUsersToDeactivateTheirOwnVirtualMFADevice",
      "Effect": "Allow",
      "Action": "iam:DeactivateMFADevice",
      "Resource": [
        "arn:aws:iam::<account>:user/${aws:username}",
        "arn:aws:iam::<account>:mfa/${aws:username}"
      ],
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    },
    {
      "Sid": "AllowUsersToDeleteTheirOwnVirtualMFADevice",
      "Effect": "Allow",
      "Action": "iam:DeleteVirtualMFADevice",
      "Resource": [
        "arn:aws:iam::<account>:user/${aws:username}",
        "arn:aws:iam::<account>:mfa/${aws:username}"
      ],
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    },
    {
      "Sid": "AllowUsersToListMFADevicesandUsersForConsole",
      "Effect": "Allow",
      "Action": [
        "iam:ListVirtualMFADevices",
        "iam:ListUsers",
        "iam:ListMFADevices"
      ],
      "Resource": "*"
    }
  ]
}
```

The other policies allow to do `sts:assumeRole` in different accounts.

**Comment by j0hnsmith** [comment]
@asiragusa from https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa_configure-api-require.html

> However such users can only assume the role if the user is authenticated using MFA

```
{
  "Version": "2012-10-17",
  "Statement": {
    "Effect": "Allow",
    "Principal": {"AWS": "ACCOUNT-B-ID"},
    "Action": "sts:AssumeRole",
    "Condition": {"Bool": {"aws:MultiFactorAuthPresent": "true"}}
  }
}
```

I read that as 'user making request must be authed with multi factor', so as I said above, after you run `aws-vault exec ...` you should be able to run the command to remove the yubikey.

So it seems the problem you have is simply you can't auth with the yubikey, not sure why not. I think you'd have to remove it manually via the console (logging in with a code from google authenticator or similar via the qr code). Basic add/remove yubikey as 2fa device is working for me, could be some other config somewhere perhaps? It was working for you before wasn't it?

**Comment by asiragusa** [comment]
The problem comes from:
```json
{
      "Sid": "AllowUsersToDeactivateTheirOwnVirtualMFADevice",
      "Effect": "Allow",
      "Action": "iam:DeactivateMFADevice",
      "Resource": [
        "arn:aws:iam::<account>:user/${aws:username}",
        "arn:aws:iam::<account>:mfa/${aws:username}"
      ],
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    },
    {
      "Sid": "AllowUsersToDeleteTheirOwnVirtualMFADevice",
      "Effect": "Allow",
      "Action": "iam:DeleteVirtualMFADevice",
      "Resource": [
        "arn:aws:iam::<account>:user/${aws:username}",
        "arn:aws:iam::<account>:mfa/${aws:username}"
      ],
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    },
```
The Condition on such rules is needed for the 2FA to be effective. If I loose an Access Key, an attacker would be able to remove the current MFA device and add a new one, making the Condition on the `sts:assumeRole` useless.

I'm not sure though why the token is not passed during the `remove-yubikey` command. If it can help here is my `~./.aws/config`

```
[profile p]
mfa_serial=arn:aws:iam::XXX:mfa/USER

[profile staging]
source_profile=p
role_arn=arn:aws:iam::YYY:role/administratorReadOnly
mfa_serial=arn:aws:iam::XXX:mfa/USER
```

**Comment by j0hnsmith** [comment]
@asiragusa I've setup my dev env again with iam users and policies etc etc, I think I see the problem...

**Comment by mtibben** [comment]
@j0hnsmith can you please rebase on master?

**Comment by j0hnsmith** [comment]
@asiragusa I've found and fixed the problem with removing a yubikey, the creds used are obtained with a TOTP. This is the policy I tested with

```
      {
         "Effect":"Deny",
         "Action":[
            "iam:DeactivateMFADevice",
            "iam:DeleteVirtualMFADevice"
         ],
         "Resource":"arn:aws:iam::${account_id}:mfa/$${aws:username}",
         "Condition":{
            "NumericGreaterThan":{
               "aws:MultiFactorAuthAge":"900"
            }
         }
      },
```

I commented out the `Contition` and was unable to remove the yubikey, got the expected error, then after uncommenting and applying the change I was able to remove the yubikey.

**Comment by j0hnsmith** [comment]
@mtibben rebased

**Comment by mtibben** [comment]
Awesome, thank you so much @j0hnsmith for this feature!

**Comment by arnuschky** [comment]
I am struggling with using Yubikey-based auth too. I built `aws-vault` just fine, following the instructions above (thanks!)

However, when I add the Yubikey to my account through the AWS console and set the MFA's arn in my `~/.aws/config`, I end up with the following:
```
Enter token for arn:aws:iam::xxxx:u2f/user/yyy/default-zzzz: cccckkahgsdfhasdlkfhaskdjf...
aws-vault: error: Failed to get credentials for xxx (source profile for xxx-yyy): ValidationError: 2 validation errors detected: Value 'cccckkahgsdfhasdlkfhaskdjf...' at 'tokenCode' failed to satisfy constraint: Member must satisfy regular expression pattern: [\d]*; Value 'cccckkahgsdfhasdlkfhaskdjf...' at 'tokenCode' failed to satisfy constraint: Member must have length less than or equal to 6
```

However, if I remove the yubikey from by account via the AWS ui and try to add it through aws-vault, I get the following:
```
aws-vault add-yubikey myuser xxx-yyy
Adding yubikey to user myuser using profile xxx-yyy
aws-vault: error: error registering yubikey
%!(EXTRA *errors.withStack=error creating virtual device: EntityAlreadyExists: MFADevice entity at the same path and name already exists.
```

I am a bit lost here. What is `add-yubikey` for actually? How can I use the yubikey? Why doesn't aws-vault accept my token?

**Comment by j0hnsmith** [comment]
`aws-vault add-yubikey` registers your yubikey as a TOTP device (not U2F) for the given username with AWS. Then when you `aws-vault exec ...` or similar it gets a 6 digit TOTP from the yubikey and uses it to create your AWS session. A token that looks like `cccckkahgsdf....` is  a U2F token, not TOTP which is 6 numbers (as you'd get from Google Authenticator or similar).

As long as you can see an entry in Yubikey Authenticator (don't _ever_ manually remove that), I think you need to

* if you've added a u2f mfa device via the interface, remove it via the interface
* ensure you have something like this in you `~/.aws/config`
```
[profile name-used-in-aws-vault]
mfa_serial=arn:aws:iam::{your account number}:mfa/{your username}
```
* run `aws-vault remove-yubikey --debug profilename username` (you can run more than once if necessary, aws seems to keep mfa devices for deleted users)
* now you should have a clean state, so can run `aws-vault add-yubikey --debug myuser xxx-yyy` (the debug should provide a little extra info)
* `aws-vault exec profilename -- aws iam list-mfa-devices --user-name your_username` or other command that requires mfa auth

**Comment by JoostSaanen** [comment]
I've got this, it seems the correct version is installed. However, `add-yubikey` is not recognized:

```
➜  ~ brew cask reinstall aws-vault
==> Satisfying dependencies
==> Downloading https://github.com/99designs/aws-vault/releases/download/v4.6.1/aws-vault-darwin-amd64
Already downloaded: /Users/joosts/Library/Caches/Homebrew/downloads/4d7070e8a60d6bf759510a251af5a7c4594f5742b9349b83a33f24eb8153c5b6--aws-vault-darwin-amd64
==> Verifying SHA-256 checksum for Cask 'aws-vault'.
==> Installing Cask aws-vault
==> Linking Binary 'aws-vault-darwin-amd64' to '/usr/local/bin/aws-vault'.
🍺  aws-vault was successfully installed!
➜  ~ aws-vault add-yubikey xxxx xxxxx-iam --touch
aws-vault: error: expected command but got "add-yubikey", try --help
```

**Comment by tlwr** [comment]
This is a super neat feature, thanks very much @j0hnsmith, and @mtibben for merging!

Sorry if I'm being impatient. Is there an ETA for a GH release of this? There are a number of aws-vault users in our org who would benefit from this feature.

Thanks again

**Comment by mtibben** [comment]
@j0hnsmith I've just run into some problems trying to create a new release of aws-vault due to the introduction of CGO and cross-compilation. I'm going to revert this from master and look to address in a new PR

**Pull request merged**